### DISCUSSION TITLE
Cannot load such file -- optparse (LoadError)

### DISCUSSION THREAD

**Discussion by neumachen** [rbenv/ruby-build#1851] [Build failures]
Hi guys, anyone encountered this issue while installing 2.7.1?

```
$ rbenv install 2.7.1
Downloading ruby-2.7.1.tar.bz2...
-> https://cache.ruby-lang.org/pub/ruby/2.7/ruby-2.7.1.tar.bz2
Installing ruby-2.7.1...

BUILD FAILED (Pop 20.04 using ruby-build 20200224-73-g0a5e059)

Inspect or clean up the working tree at /tmp/ruby-build.20200821113417.841610.nXIFsy
Results logged to /tmp/ruby-build.20200821113417.841610.log

Last 10 log lines:
        LC_CTYPE =
        MFLAGS =
gcc (Ubuntu 9.3.0-10ubuntu2) 9.3.0
Copyright (C) 2019 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

./tool/file2lastrev.rb:6:in `require': cannot load such file -- optparse (LoadError)
        from ./tool/file2lastrev.rb:6:in `<main>'
make: *** [uncommon.mk:1130: revision.tmp] Error 1
```

**Comment by hamidnazari** [comment]
Nope!

```
rbenv install 2.7.1
Downloading ruby-2.7.1.tar.bz2...
-> https://cache.ruby-lang.org/pub/ruby/2.7/ruby-2.7.1.tar.bz2
Installing ruby-2.7.1...
ruby-build: using readline from homebrew
Installed ruby-2.7.1 to /Users/hamid/.rbenv/versions/2.7.1
```

**Comment by mlue** [comment]
fixed this by uninstall rbenv and reinstall with homebrew for ubuntu

**Comment by gencer** [comment]
10 days ago I've installed `rbenv` and `ruby-2.7.1` with success as usual. Today, side-by-side I wanted to install `ruby-2.7.2` and same error occured.

```
[09:51] [gencer] [bf ~/.rbenv] $ rbenv install 2.7.2
Downloading ruby-2.7.2.tar.bz2...
-> https://cache.ruby-lang.org/pub/ruby/2.7/ruby-2.7.2.tar.bz2
Installing ruby-2.7.2...

BUILD FAILED (Ubuntu 20.04 using ruby-build 20201118-6-g4a0c87c)

Inspect or clean up the working tree at /tmp/ruby-build.20201206095154.809205.OWRJt4
Results logged to /tmp/ruby-build.20201206095154.809205.log

Last 10 log lines:
linking shared-object ripper.so
make[2]: Leaving directory '/tmp/ruby-build.20201206095154.809205.OWRJt4/ruby-2.7.2/ext/ripper'
make[2]: Entering directory '/tmp/ruby-build.20201206095154.809205.OWRJt4/ruby-2.7.2'
./tool/file2lastrev.rb:6:in `require': cannot load such file -- optparse (LoadError)
        from ./tool/file2lastrev.rb:6:in `<main>'
make[2]: *** [uncommon.mk:1130: revision.tmp] Error 1
make[2]: Leaving directory '/tmp/ruby-build.20201206095154.809205.OWRJt4/ruby-2.7.2'
make[1]: *** [exts.mk:86: ruby] Error 2
make[1]: Leaving directory '/tmp/ruby-build.20201206095154.809205.OWRJt4/ruby-2.7.2'
make: *** [uncommon.mk:295: build-ext] Error 2
```

I've pulled all `rbenv` + `ruby-build` changes. Not worked.

Manually uninstalling rbenv and reinstalling 2.7.1 and 2.7.2 solved my problem.

**Comment by yahonda** [comment]
[ruby-build.20210407222345.480785.log](https://github.com/rbenv/ruby-build/files/6271964/ruby-build.20210407222345.480785.log)
Looks like I have managed to reproduce this issue. Clean install rbenv and ruby-build then install 3.0.0 and 3.0.1.

Here are steps to reproduce on Ubuntu 20.10 x86_64

- Remove ~/.rbenv and install rbenv by "Basic GitHub Checkout" method as follows.

```
$ rm -rf .rbenv/
$ git clone https://github.com/rbenv/rbenv.git ~/.rbenv
Cloning into '/home/yahonda/.rbenv'...
remote: Enumerating objects: 2886, done.
remote: Total 2886 (delta 0), reused 0 (delta 0), pack-reused 2886
Receiving objects: 100% (2886/2886), 562.66 KiB | 2.08 MiB/s, done.
Resolving deltas: 100% (1801/1801), done.
$ cd ~/.rbenv && src/configure && make -C src
make: Entering directory '/home/yahonda/.rbenv/src'
gcc -fPIC     -c -o realpath.o realpath.c
gcc -shared -Wl,-soname,../libexec/rbenv-realpath.dylib  -o ../libexec/rbenv-realpath.dylib realpath.o
make: Leaving directory '/home/yahonda/.rbenv/src'
$ echo 'export PATH="$HOME/.rbenv/bin:$PATH"' >> ~/.profile
$ ~/.rbenv/bin/rbenv init
# Load rbenv automatically by appending
# the following to ~/.bashrc:

eval "$(rbenv init -)"
$ vi ~/.bashrc # to add `eval "$(rbenv init -)"`
$ exec $SHELL -l
$ curl -fsSL https://github.com/rbenv/rbenv-installer/raw/master/bin/rbenv-doctor | bash
Checking for `rbenv' in PATH: /home/yahonda/.rbenv/bin/rbenv
Checking for rbenv shims in PATH: OK
Checking `rbenv install' support: not found
Unless you plan to add Ruby versions manually, you should install ruby-build.
Please refer to https://github.com/rbenv/ruby-build#installation

Counting installed Ruby versions: none
  There aren't any Ruby versions installed under `/home/yahonda/.rbenv/versions'.
  You can install Ruby versions like so: rbenv install 2.2.4
Checking RubyGems settings: OK
Auditing installed plugins: OK
```

- Install ruby-build as "rbenv plugin" and install Ruby 3.0.0

```
$ mkdir -p "$(rbenv root)"/plugins
$ git clone https://github.com/rbenv/ruby-build.git "$(rbenv root)"/plugins/ruby-build
Cloning into '/home/yahonda/.rbenv/plugins/ruby-build'...
remote: Enumerating objects: 162, done.
remote: Counting objects: 100% (162/162), done.
remote: Compressing objects: 100% (114/114), done.
remote: Total 11469 (delta 115), reused 73 (delta 39), pack-reused 11307
Receiving objects: 100% (11469/11469), 2.43 MiB | 4.02 MiB/s, done.
Resolving deltas: 100% (7582/7582), done.
$ rbenv install 3.0.0
Downloading ruby-3.0.0.tar.gz...
-> https://cache.ruby-lang.org/pub/ruby/3.0/ruby-3.0.0.tar.gz
Installing ruby-3.0.0...
Installed ruby-3.0.0 to /home/yahonda/.rbenv/versions/3.0.0

$ rbenv global 3.0.0
$ which ruby
/home/yahonda/.rbenv/shims/ruby
$ ruby -v
ruby 3.0.0p0 (2020-12-25 revision 95aff21468) [x86_64-linux]
```

- Install Ruby 3.0.1 and it gets the error below

```
$ rbenv install 3.0.1
Downloading ruby-3.0.1.tar.gz...
-> https://cache.ruby-lang.org/pub/ruby/3.0/ruby-3.0.1.tar.gz
Installing ruby-3.0.1...

BUILD FAILED (Ubuntu 20.10 using ruby-build 20210405-2-g693a28e)

Inspect or clean up the working tree at /tmp/ruby-build.20210407222345.480785.HSbVgV
Results logged to /tmp/ruby-build.20210407222345.480785.log

Last 10 log lines:
linking shared-object ripper.so
make[2]: Leaving directory '/tmp/ruby-build.20210407222345.480785.HSbVgV/ruby-3.0.1/ext/ripper'
make[2]: Entering directory '/tmp/ruby-build.20210407222345.480785.HSbVgV/ruby-3.0.1'
./tool/file2lastrev.rb:6:in `require': cannot load such file -- optparse (LoadError)
	from ./tool/file2lastrev.rb:6:in `<main>'
make[2]: *** [uncommon.mk:1156: revision.tmp] Error 1
make[2]: Leaving directory '/tmp/ruby-build.20210407222345.480785.HSbVgV/ruby-3.0.1'
make[1]: *** [exts.mk:88: ruby] Error 2
make[1]: Leaving directory '/tmp/ruby-build.20210407222345.480785.HSbVgV/ruby-3.0.1'
make: *** [uncommon.mk:300: build-ext] Error 2
$
```

**Comment by yahonda** [comment]
It does not reproduce on macOS 11.3 Beta (20E5224a). I do not have other non-beta versions of macOS.

**Comment by AlexWayfer** [comment]
The same for me with Ruby 2.7.3 version. 3.0.1, 2.6.7 and 2.5.9 installed without errors.

The second try to install is successful somewhy.

**Comment by smoothdvd** [comment]
@yahonda same error here on Ubuntu 20.04.2 LTS.

```
`require': cannot load such file -- optparse
```

Look script cannot load optparse.

**Comment by mislav** [comment]
Such a bizarre error. I cannot reproduce. Those who can: do you have RUBYOPT or RUBYLIB set in your environment variables, by any chance?

**Comment by smoothdvd** [comment]
@mislav No such environment variables.

**Comment by smoothdvd** [comment]
@mislav Finally I install 3.0.1 success as following step

1. Install 2.7.3
2. Set global version to 2.7.3
3. Install 3.0.1 again, no error.

**Comment by yahonda** [comment]
None of these environment variables configured.

```
$ echo $RUBYOPT

$ echo $RUBYLIB

$
```

**Comment by mislav** [comment]
@hsbt Do you know if building a Ruby version is dependent or at all affected by having another Ruby version installed on the system? The reproduction steps so far suggest that the LoadError pops up depending on which Ruby version is being installed and which other Ruby version was selected as "global".

**Comment by hsbt** [comment]
https://github.com/rbenv/ruby-build/issues/1484#issuecomment-814918917 is a strange behavior for me. Because I never faced that with my macOS and linux boxes.

I try to reproduce it.

**Comment by andreaslillebo** [comment]
I ran into the same error installing 3.0.1 with asdf on Ubuntu 20.04.

Following similar steps to @smoothdvd, I was able to install 3.0.1:
1. Switching the global version from 3.0.0 to 2.7.2 with `asdf global ruby 2.7.2`
2. Install ruby 3.0.1 with `asdf install ruby 3.0.1`

**Comment by andremmfaria** [comment]
@mislav , You are correct on our comment. This hapenned to me from version 2.6.8 to version 2.7.3. I solved it as follows:

I am using CentOS 7 on a corporate environment. For that reason my options are quite limited. 
Disable or remove __ALL__ other `ruby` and `gem` installations from your environment. I did that by starting a plain zsh session with just the standard $PATH environment variables loaded. 

1. Install rbenv

```
$ git clone https://github.com/rbenv/rbenv.git ~/.rbenv
$ cd ~/.rbenv
$ src/configure && make -C src
```

2. Add `$HOME/.rbenv/bin` to your `$PATH`

3. Install ruby-build
```
$ git clone https://github.com/rbenv/ruby-build.git ~/ruby-build
$ cd ~/ruby-build
$ sudo PREFIX=/usr/local/ ./install.sh
```

4. Ensure `/usr/local/bin` is in your `$PATH`.

5. Install version `2.6.8` and set it as global
```
$ sudo yum install -y openssl-devel readline-devel
$ rbenv install 2.6.8
$ rbenv global 2.6.8
```
You can check if it is correct using this command:
```
$ rbenv versions
  system
  2.6.8 (set by ~/.rbenv/version)
```

6. Install optparse using this newly installed version:
```
$ ~/.rbenv/versions/2.6.8/bin/gem install optparse
```

7. Install version `2.7.3` and set it as global:
```
$ rbenv install 2.7.3
$ rbenv global 2.7.3
```
You can check if it is correct using this command:
```
$ rbenv versions
  system
  2.6.8
* 2.7.3 (set by ~/.rbenv/version)
```

I believe this issue started from version 2.6.x to version 2.7.x as to update from 2.5.x to 2.6.x i did not have this issue.
I believe that the same steps need to be made from 2.6.x to 3.x

**Comment by nihilismus** [comment]
I'm facing the same issue with `2.7.0` in `Void Linux x86_64`.

My solution, based on previous comments, was to set `RUBY_CONFIGURE_OPTS` to indicate the _"BASERUBY"_ (which I don't know what it means) and then install as usual:

```shell
$ rbenv versions
* system
  2.5.7
  2.6.6
$ export RUBY_CONFIGURE_OPTS=" --with-baseruby=$HOME/.rbenv/versions/2.6.6/bin/ruby"
$ rbenv install 2.7.0
Installing ruby-2.7.0...
Installed ruby-2.7.0 to $HOME/.rbenv/versions/2.7.0
```

---

From the log, when `rbenv install 2.7.0` fails, I can see the next message:

```
   * BASERUBY -v:         ruby 2.7.4p191 (2021-07-07 revision a21a3b7d23) \
                          [x86_64-linux]

---
	BASERUBY = $HOME/.rbenv/shims/ruby --disable=gems
```

And notice that `ruby 2.7.4p191` is _system_:

```
$ rbenv shell system
$ ruby --version
ruby 2.7.4p191 (2021-07-07 revision a21a3b7d23) [x86_64-linux]
```

Now, reading <https://github.com/ruby/ruby/blob/master/configure.ac#L50> makes me wonder if a _previous_ version of Ruby is needed to build Ruby `>= 2.7.0`?: _"executable host ruby is required.  use `--with-baseruby` option"_.

Whats your opinion? 🤔

**Comment by mislav** [comment]
> makes me wonder if a _previous_ version of Ruby is needed to build Ruby `>= 2.7.0`?: _"executable host ruby is required. use `--with-baseruby` option"_.

@hsbt Do you know anything about this?

**Comment by mcandre** [comment]
Same problem on my machine. Unable to install Ruby 2.7.2 via rvm. I have 2.7.0.

**Comment by sanemat** [comment]
I met same issue on Ubuntu 21.04 amd64.

I used system ruby instead of ruby-build's ruby :sweat_smile:  

```
sudo apt install ruby
rbenv global system
rbenv rehash
rbenv install 2.7.4
```

refs: [ruby-buildでRubyをインストールしようとしたらoptparseが無いって言われる - ser1zw's blog](https://ser1zw.hatenablog.com/entry/2021/07/11/034730)

**Comment by zw963** [comment]
This issue happen on, when install new ruby version use RVM, but current, selected version is system ruby.

that is, this is a issue come from system ruby, don't use it, you can just remove system ruby use package manager, or switch to use a more older version ruby which installed use RVM.
